examples/vision/patch_conv_net.py

```diff
@@ -0,0 +1,661 @@
+
+The patch-conv-net is shown in the figure below.
+
+| ![image model](https://i.imgur.com/NHiQeac.png) |
+| :--: |
+| [Source](https://arxiv.org/abs/2112.13692) |
+
+All the modules in the architecture are built in the earlier seciton.
+In this section, we will stack all of the different modules together.
+"""
+
+
+class PatchConvNet(keras.Model):
+    def __init__(
+        self,
+        stem,
+        trunk,
+        attention_pooling,
+        train_augmentation_model,
+        test_augmentation_model,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.stem = stem
+        self.trunk = trunk
+        self.attention_pooling = attention_pooling
+        self.train_augmentation_model = train_augmentation_model
+        self.test_augmentation_model = test_augmentation_model
+
+    def get_config(self):
+        config = super().get_config()
+        config.update(
+            {
+                "stem": self.stem,
+                "trunk": self.trunk,
+                "attention_pooling": self.attention_pooling,
+                "train_augmentation_model": self.train_augmentation_model,
+                "test_augmentation_model": self.test_augmentation_model,
+            }
+        )
+        return config
+
+    def calculate_loss(self, inputs, test=False):
+        images, labels = inputs
+        # Augment the input images.
+        if test:
+            augmented_images = self.test_augmentation_model(images)
+        else:
+            augmented_images = self.train_augmentation_model(images)
+        # Pass through the stem.
+        x = self.stem(augmented_images)
+        # Pass through the trunk.
+        x = self.trunk(x)
+        # Pass through the attention pooling block.
+        logits, _ = self.attention_pooling(x)
+        # Compute the total loss.
+        total_loss = self.compiled_loss(labels, logits)
+        return total_loss, logits
+
+    def train_step(self, inputs):
+        with tf.GradientTape() as tape:
+            total_loss, logits = self.calculate_loss(inputs)
+        # Apply gradients.
+        train_vars = [
+            self.stem.trainable_variables,
+            self.trunk.trainable_variables,
+            self.attention_pooling.trainable_variables,
+        ]
+        grads = tape.gradient(total_loss, train_vars)
+        tv_list = []
+        for (grad, var) in zip(grads, train_vars):
+            for g, v in zip(grad, var):
+                tv_list.append((g, v))
+        self.optimizer.apply_gradients(tv_list)
+        # Report progress.
+        _, labels = inputs
+        self.compiled_metrics.update_state(labels, logits)
+        return {m.name: m.result() for m in self.metrics}
+
+    def test_step(self, inputs):
+        total_loss, logits = self.calculate_loss(inputs, test=True)
+        # Report progress.
+        _, labels = inputs
+        self.compiled_metrics.update_state(labels, logits)
+        return {m.name: m.result() for m in self.metrics}
+
+    @tf.function(
```

Comment:
You don't need to add a tf.function on `call`, typically. You train with `fit()` so this is done automatically.